processing/src/main/java/org/apache/druid/query/context/ResponseContext.java

```diff
@@ -129,214 +349,389 @@
+     * It is not updated continuously as {@link Keys#TIMEOUT_AT}.
      */
-    QUERY_TOTAL_BYTES_GATHERED("queryTotalBytesGathered"),
+    public static final Key QUERY_FAIL_DEADLINE_MILLIS = new LongKey(
+        "queryFailTime",
+        false);
+
     /**
      * This variable indicates when a running query should be expired,
      * and is effective only when 'timeout' of queryContext has a positive value.
      * Continuously updated by {@link org.apache.druid.query.scan.ScanQueryEngine}
      * by reducing its value on the time of every scan iteration.
      */
-    TIMEOUT_AT("timeoutAt"),
+    public static final Key TIMEOUT_AT = new LongKey(
+        "timeoutAt",
+        false);
+
     /**
-     * The number of scanned rows.
-     * For backward compatibility the context key name still equals to "count".
+     * The number of rows scanned by {@link org.apache.druid.query.scan.ScanQueryEngine}.
+     *
+     * Named "count" for backwards compatibility with older data servers that still send this, even though it's now
+     * marked as internal.
      */
-    NUM_SCANNED_ROWS(
+    public static final Key NUM_SCANNED_ROWS = new CounterKey(
         "count",
-            (oldValue, newValue) -> ((Number) oldValue).longValue() + ((Number) newValue).longValue()
-    ),
+        false);
+
     /**
      * The total CPU time for threads related to Sequence processing of the query.
      * Resulting value on a Broker is a sum of downstream values from historicals / realtime nodes.
      * For additional information see {@link org.apache.druid.query.CPUTimeMetricQueryRunner}
      */
-    CPU_CONSUMED_NANOS(
+    public static final Key CPU_CONSUMED_NANOS = new CounterKey(
         "cpuConsumed",
-            (oldValue, newValue) -> ((Number) oldValue).longValue() + ((Number) newValue).longValue()
-    ),
+        false);
+
     /**
      * Indicates if a {@link ResponseContext} was truncated during serialization.
      */
-    TRUNCATED(
+    public static final Key TRUNCATED = new BooleanKey(
         "truncated",
-            (oldValue, newValue) -> (boolean) oldValue || (boolean) newValue
-    );
+        false);
+
+    /**
+     * One and only global list of keys. This is a semi-constant: it is mutable
+     * at start-up time, but then is not thread-safe, and must remain unchanged
+     * for the duration of the server run.
+     */
+    public static final Keys INSTANCE = new Keys();
 
     /**
      * ConcurrentSkipListMap is used to have the natural ordering of its keys.
-     * Thread-safe structure is required since there is no guarantee that {@link #registerKey(BaseKey)}
+     * Thread-safe structure is required since there is no guarantee that {@link #registerKey(Key)}
      * would be called only from class static blocks.
      */
-    private static final ConcurrentMap<String, BaseKey> REGISTERED_KEYS = new ConcurrentSkipListMap<>();
+    private final ConcurrentMap<String, Key> registered_keys = new ConcurrentSkipListMap<>();
```

Comment:
```suggestion
    private final ConcurrentMap<String, Key> registeredKeys = new ConcurrentSkipListMap<>();
```
```suggestion
    private final ConcurrentMap<String, Key> registered_keys = new ConcurrentSkipListMap<>();
```

Reply:
@jihoonson, thanks for catching that one! Renamed as suggested.